### PR TITLE
Support for shared modules

### PR DESCRIPTION
**Pull request by jpakkane**
With a test to verify that `-Wl,--no-undefined` is not used. May be a bit rough still:

 - not tested on Windows, not sure if MSVC wants extra flags
 - left prefix at `lib` instead of empty because both are used in the wild and the former is less surprising

### CODE REVIEW

**Comment by QuLogic** [comment]
The correct arguments are probably in #834.

**Comment by jpakkane** [comment]
Apparently leaving unresolved symbols [is impossible in OSX](https://developer.apple.com/library/content/documentation/Porting/Conceptual/PortingUnix/compiling/compiling.html#//apple_ref/doc/uid/TP40002850-SW1).

> If the compiler finds symbol requests in the plugin that cannot be resolved in the application, you will get a link error. This means that you must use the -l flag to link against any libraries that the plugin requires as though you were building a complete executable.

> Important: OS X does not support the concept of weak linking as it is found in systems like Linux. If you override one symbol, you must override all of the symbols in that object file.

**Comment by nirbheek** [comment]
It is definitely possible on OS X to link an executable with symbols that are unresolved at build time, but might be provided at run time. This is exactly what caused all the XCode 8 breakage that that I added [a PR](https://github.com/mesonbuild/meson/pull/949) about.

It is also possible on Windows since I have done it in the past.

**Comment by jpakkane** [comment]
This branch has it working for Windows (but not tested as per the comment in the executable source). For OSX, I just go with what Apple's documentation tells me. If someone knows the magic arguments to use in this case, please let me know.

**Comment by QuLogic** [comment]
The undefined symbol test in #834 should work on all platforms.

**Comment by nirbheek** [comment]
Yep, we already went through all the research needed to make it work on all platforms.

**Comment by jpakkane** [comment]
Thanks, this should now be up to date and work on all platforms.

**Comment by hughsie** [comment]
It works great for me, thanks!

File: test cases/common/125 shared module/module.c [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  #endif
+#endif
+
+/*
+ * Shared modules often have references to symbols that are not defined
+ * at link time, but which will be provided from deps of the executable that
+ * dlopens it. We need to make sure that this works, i.e. that we do
+ * not pass -Wl,--no-undefined when linking modules.
+ */
+int func_from_language_runtime();
```
Comment (QuLogic) [inline test cases/common/125 shared module/module.c]: Missing `void` to be a valid C prototype.

File: test cases/common/125 shared module/module.c [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+int DLL_PUBLIC func() {
```
Comment (QuLogic) [inline test cases/common/125 shared module/module.c]: As is this one.

File: test cases/common/125 shared module/prog.c [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    return 0;
+}
+
+#else
+
+#include<dlfcn.h>
+#include<assert.h>
+#include<stdio.h>
+
+int func();
```
Comment (QuLogic) [inline test cases/common/125 shared module/prog.c]: Again, both missing `void`.

File: test cases/common/125 shared module/prog.c [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+int func_from_language_runtime();
+
+int main(int argc, char **argv) {
+    void *dl;
+    int (*importedfunc)();
```
Comment (QuLogic) [inline test cases/common/125 shared module/prog.c]: Missing `void` as well.

File: test cases/common/125 shared module/runtime.c [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    #define DLL_PUBLIC
+  #endif
+#endif
+
+/*
+ * This file pretends to be a language runtime that supports extension
+ * modules.
+ */
+
+int DLL_PUBLIC func_from_language_runtime() {
```
Comment (QuLogic) [inline test cases/common/125 shared module/runtime.c]: Missing `void`.

**Comment by QuLogic** [comment]
On the failing build, it's:
```
Link:

  c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\bin\link.exe /ERRORREPORT:QUEUE /OUT:".\mymodule.dll" /NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib /MANIFEST /ManifestFile:"mymodule@sha\mymodule.dll.intermediate.manifest" /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /PDB:"C:\projects\meson\b 74ntnd6v\mymodule.pdb" /SUBSYSTEM:WINDOWS /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"mymodule.lib" /MACHINE:X86 /DLL "mymodule@sha\module.obj" /DLL /FORCE:UNRESOLVED 

module.obj : warning LNK4075: ignoring '/EDITANDCONTINUE' due to '/OPT:ICF' specification [C:\projects\meson\b 74ntnd6v\mymodule@sha.vcxproj]

     Creating library mymodule.lib and object mymodule.exp

module.obj : error LNK2019: unresolved external symbol _func_from_language_runtime referenced in function _func [C:\projects\meson\b 74ntnd6v\mymodule@sha.vcxproj]

.\mymodule.dll : warning LNK4088: image being generated due to /FORCE option; image may not run [C:\projects\meson\b 74ntnd6v\mymodule@sha.vcxproj]
```

It prints both an error _and_ a warning that it's going to continue. I wonder if it's a bug in `msbuild`? Because the same output occurs in #834 and it passed...

**Comment by nirbheek** [comment]
Turns out that we can't actually create modules like that on Windows. I tested it and it just doesn't work. What you need to do is use `CreateToolhelp32Snapshot ` + `Module32First` + `GetProcAddress` + `Module32Next`. The test now does this and we no longer pass `/FORCE:UNRESOLVED` with MSVC.

Incidentally, this also fixes the test on MinGW since MinGW also doesn't support unresolved symbols in DLLs for this very reason.

PS: This is exactly what gmodule does in glib.

File: test cases/common/125 shared module/module.c [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            if ((f = (fptr) GetProcAddress (me32.hModule, name)) != NULL)
+                break;
+        } while (Module32Next (snapshot, &me32));
+    }
+
+    CloseHandle (snapshot);
+    return f;
+}
+
+int DLL_PUBLIC func() {
```
Comment (QuLogic) [inline test cases/common/125 shared module/module.c]: Missing `void` here too.

File: test cases/common/125 shared module/module.c (hunk 1 above)
Comment (QuLogic) [inline test cases/common/125 shared module/module.c]: Still missing.

File: test cases/common/125 shared module/prog.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
+
+#include <stdio.h>
+
+int func_from_language_runtime();
```
Comment (QuLogic) [inline test cases/common/125 shared module/prog.c]: `void` missing here.

File: test cases/common/125 shared module/prog.c [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                    NULL, GetLastError (), 0,
+                    (LPWSTR) &msg, 0, NULL);
+    return msg;
+}
+
+int
+main (int argc, char **argv)
+{
+    HINSTANCE handle;
+    int (*importedfunc)();
```
Comment (QuLogic) [inline test cases/common/125 shared module/prog.c]: Probably `void` needed here.

File: test cases/common/125 shared module/prog.c [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+#else
+
+#include<dlfcn.h>
+#include<assert.h>
+
+int main(int argc, char **argv) {
+    void *dl;
+    int (*importedfunc)();
```
Comment (QuLogic) [inline test cases/common/125 shared module/prog.c]: And probably here too.

**Comment by nirbheek** [comment]
All specified instances of missing `void` fixed.

**Comment by arteymix** [comment]
It's pretty common practice with Vala to declare shared modules and I wondered if I could add a test case (via patch or anything) to this PR.

**Comment by nirbheek** [comment]
I don't think Jussi will mind. Go for it! :)

**Comment by jpakkane** [comment]
I'll merge this now. Please file a new MR with the test. Thanks.

**Pull request merged**